include/ginkgo/core/base/machine_topology.hpp

```diff
@@ -0,0 +1,420 @@
+     *
+     * @param id  The id of the pci device
+     * @return  the PCI object struct.
+     */
+    const io_obj_info *get_pci_device(size_type id) const
+    {
+        GKO_ENSURE_IN_BOUNDS(id, this->pci_devices_.size());
+        return &this->pci_devices_[id];
+    }
+
+    /**
+     * Get the object of type pci device associated with the PCI bus id.
+     *
+     * @param pci_bus_id  The PCI bus id of the pci device
+     * @return  the PCI object struct.
+     */
+    const io_obj_info *get_pci_device(const std::string &pci_bus_id) const;
+
+    /**
+     * Get the number of PU objects stored in this Topology tree.
+     *
+     * @return  the number of PUs.
+     */
+    size_type get_num_pus() const { return this->pus_.size(); }
+
+    /**
+     * Get the number of core objects stored in this Topology tree.
+     *
+     * @return  the number of cores.
+     */
+    size_type get_num_cores() const { return this->cores_.size(); }
+
+    /**
+     * Get the number of PCI device objects stored in this Topology tree.
+     *
+     * @return  the number of PCI devices.
+     */
+    size_type get_num_pci_devices() const { return this->pci_devices_.size(); }
+
+    /**
+     * Get the number of NUMA objects stored in this Topology tree.
+     *
+     * @return  the number of NUMA objects.
+     */
+    size_type get_num_numas() const { return this->num_numas_; }
+
+protected:
+    /**
+     * Creates a new MachineTopology object.
+     */
+    static std::shared_ptr<MachineTopology> create()
+    {
+        return std::shared_ptr<MachineTopology>(new MachineTopology());
+    }
+
+    /**
+     * The default constructor that loads the different objects.
+     */
+    MachineTopology();
+
+    /**
+     * @internal
+     *
+     * A helper function that binds the object with an id.
+     */
+    void hwloc_binding_helper(
+        const std::vector<MachineTopology::normal_obj_info> &obj, const int *id,
+        const size_type num_ids) const;
+
+    /**
+     * @internal
+     *
+     * Load the objects of a normal HWLOC type (Packages, cores, numa-nodes).
+     *
+     * @note The objects should be sorted by logical index since hwloc uses
+     * logical index with these functions
+     */
+    void load_objects(hwloc_obj_type_t type,
+                      std::vector<normal_obj_info> &objects) const;
+
+    /**
+     * @internal
+     *
+     * Load the objects of io type (PCI devices and OS devices).
+     *
+     * @note The objects should be sorted by logical index since hwloc uses
+     * logical index with these functions
+     */
+    void load_objects(hwloc_obj_type_t type,
+                      std::vector<io_obj_info> &vector) const;
+
+    /**
+     * @internal
+     *
+     * Initialize the topology object.
+     */
+    hwloc_topology *init_topology();
+
+    /**
+     * @internal
+     *
+     * Initialize the bitmap.
+     */
+    hwloc_bitmap_s *init_bitmap();
```

Comment:
These should not return plain (unmanaged) pointers since it would be prone to memory leaks, hard to use correctly / maintain and not thread safe.